src/idlcxx/src/traits.c

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright(c) 2020 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <string.h>
+
+#include "idl/stream.h"
+#include "idl/processor.h"
+
+#include "generator.h"
+
+static idl_retcode_t
+emit_traits(
+  const idl_pstate_t* pstate,
+  const bool revisit,
+  const idl_path_t* path,
+  const void* node,
+  void* user_data)
+{
+  struct generator *gen = user_data;
+  char *name;
+  const char *fmt, *keyless = "true";
+  const idl_struct_t *_struct = node;
+
+  (void)pstate;
+  (void)revisit;
+  (void)path;
+
+  fmt = "template <>\n"
+        "class TopicTraits<%1$s>\n"
+        "{\n"
+        "public:\n"
+        "  static ::org::eclipse::cyclonedds::topic::DataRepresentationId_t getDataRepresentationId()\n"
+        "  {\n"
+        "    return ::org::eclipse::cyclonedds::topic::OSPL_REPRESENTATION;\n"
+        "  }\n\n"
+        "  static ::std::vector<uint8_t> getMetaData()\n"
+        "  {\n"
+        "    return ::std::vector<uint8_t>();\n"
+        "  }\n\n"
+        "  static ::std::vector<uint8_t> getTypeHash()\n"
+        "  {\n"
+        "    return ::std::vector<uint8_t>();\n"
+        "  }\n\n"
+        "  static ::std::vector<uint8_t> getExtentions()\n"
+        "  {\n"
+        "    return ::std::vector<uint8_t>();\n"
+        "  }\n\n"
+        "  static bool isKeyless()\n"
+        "  {\n"
+        "    return %3$s;\n"
+        "  }\n\n"
+        "  static const char *getTypeName()\n"
+        "  {\n"
+        "    return \"%2$s\";\n" /* skip preceeding "::" according to convention */
+        "  }\n\n"
+        "  static ddsi_sertopic *getSerTopic(const std::string& topic_name)\n"
+        "  {\n"
+        "    auto *st = new ddscxx_sertopic<%1$s>(topic_name.c_str());\n"
```

Comment:
I don't see the definitions of `ddsi_sertopic`, `ddscxx_sertopic` getting included in the generated code, and they are not either part of `TopicTraits.hpp` or `DataRepresentation.hpp`, how are these included?